uast/uast-encoding-v2.md

```diff
@@ -0,0 +1,249 @@
+
+Value nodes may be deduplicated when encoding the UAST, thus implementations should make a
+copy of the value when decoding.
+
+#### Array
+
+Array nodes must only have an `id` (optional), `values` and `values_offs` fields set.
+`is_object` field may also be set to `false` when encoding an empty array.
+
+```proto
+message Node {
+    uint64 id = 1;
+    
+    repeated uint64 values = 8;
+    uint64 values_offs = 11;
+    
+    bool is_object = 9;
+}
+```
+
+`values` field contains a list of node IDs of the children nodes in the array. The list order
+corresponds to an array element order. Nil elements are encoded as ID value `0`.
+
+`values_offs` is an optional node ID offset that should be added to all `values` elements
+before resolving that child ID. Used for compression.
+
+Empty array are encoded by setting `id` field (optional) and optionally setting `is_object`
+to `false`. Message with no fields set is a valid empty array. `values` may be also set to
+an empty array and/or the `values_offs` may be set to `0`.
+
+#### Object
+
+Object nodes are represented by two arrays: one for object keys (`keys` and `keys_from`)
+and the second one for values (`values` and `values_offs`).
+
+```proto
+message Node {
+    uint64 id = 1;
+    
+    repeated uint64 keys = 7;
+    uint64 keys_from = 10;
+    
+    repeated uint64 values = 8;
+    
+    bool is_object = 9;
+    
+    uint64 values_offs = 11;
+}
+```
+
+Empty objects must have a `is_object` fields set. Other fields may be either set to zero value,
+or must not be set at all.
+
+Keys array can be either specified explicitly with `keys`, or copied from another object
+by specifying its ID in `keys_from`. Both `keys` and `keys_from` must not be set.
+
+Values are decoded the same way as for [Arrays](#array). The size of values array must be
+the same as arrays of keys.
+
+Objects are considered to be a unordered sets, thus keys and values may be reordered for efficiency.
+
+### Tree reconstruction
+
+The state needed to decode the file consist of a `current_id` integer (set to `1` initially)
+used for automatic `id` field generation when it's not set and a map of node IDs to [Node](#node)
+structures.
+
+It must also keep a set of seen nodes (`seen`) when reconstructing the tree to prevent loops.
+It must also keep a flag for each entry to this set to track if a given node is a tree branch,
+as opposed to a direct acyclic graph (DAG) branch. Nodes in tree branches are referenced
+only once, as opposed to DAGs where branches may overlap.
+
+Decoding process is can be the following:
+
+- The decoder read each node message from the file.
+- It assigns an `id` automatically from `current_id` and then incrementing the `current_id`,
+  or sets `current_id` to `id + 1` if the `id` field was set in the message.
+- It must ensure that `id` values are monotonically increasing, including both auto-assigned
```

Comment:
What if they are not? Should the reader fixup and backpatch the id sequence or report an error?